client/service/src/chain_ops.rs

```diff
@@ -116,85 +362,86 @@ impl<
 		let import = future::poll_fn(move |cx| {
 			let client = &self.client;
 			let queue = &mut self.import_queue;
-
-			// Start by reading the number of blocks if not done so already.
-			let count = match count {
-				Some(c) => c,
-				None => {
-					let c: u64 = match Decode::decode(&mut io_reader_input) {
-						Ok(c) => c,
-						Err(err) => {
-							let err = format!("Error reading file: {}", err);
-							return std::task::Poll::Ready(Err(From::from(err)));
+			match state.take().expect("state should never be None; qed") {
+				ImportState::Reading{mut block_iter} => {
+					match block_iter.next() {
+						None => {
+							// The iterator is over: we now need to wait for the import queue to finish.
+							let num_expected_blocks = block_iter.num_expected_blocks();
+							let read_block_count = block_iter.read_block_count();
+							let delay = Duration::from_millis(DELAY_TIME);
+							state = Some(ImportState::WaitingForImportQueueToFinish{num_expected_blocks, read_block_count, delay});
 						},
-					};
-					info!("📦 Importing {} blocks", c);
-					count = Some(c);
-					c
-				}
-			};
-
-			// Read blocks from the input.
-			if read_block_count < count {
-				match SignedBlock::<Self::Block>::decode(&mut io_reader_input) {
-					Ok(signed) => {
-						let (header, extrinsics) = signed.block.deconstruct();
-						let hash = header.hash();
-						// import queue handles verification and importing it into the client
-						queue.import_blocks(BlockOrigin::File, vec![
-							IncomingBlock::<Self::Block> {
-								hash,
-								header: Some(header),
-								body: Some(extrinsics),
-								justification: signed.justification,
-								origin: None,
-								allow_missing_state: false,
-								import_existing: force,
+						Some(block_result) => {
+							let read_block_count = block_iter.read_block_count();
+							match block_result {
+								Ok(block) => {
+									if read_block_count - link.imported_blocks >= MAX_PENDING_BLOCKS {
+										// The queue is full, so do not add this block and simply wait until
+										// the queue has made some progress.
+										let delay = Duration::from_millis(DELAY_TIME);
+										state = Some(ImportState::WaitingForImportQueueToCatchUp{block_iter, delay, block});
+									} else {
+										// Queue is not full, we can keep on adding blocks to the queue.
+										import_block_to_queue(block, queue, force);
+										state = Some(ImportState::Reading{block_iter});
+									}
+								}
+								Err(e) => {
+									return std::task::Poll::Ready(
+										Err(Error::Other(format!("Error reading block #{}: {}", read_block_count, e))))
+								}
 							}
-						]);
+						}
+					}
+				},
+				ImportState::WaitingForImportQueueToCatchUp{block_iter, delay, block} => {
+					let read_block_count = block_iter.read_block_count();
+					if read_block_count - link.imported_blocks >= MAX_PENDING_BLOCKS {
+						thread::sleep(delay);
```

Comment:
Is this the correct way of doing this? The goal here is to not call the waker right now, but rather say "you can wait for `delay` time because we know the queue won't be empty by that time"...

Reply:
Sleeping is generally not a good way to do async IO. Here you could simply return `Pending` and organize this function to be polled when the queue is done processing. The queue signals its progress though `WaitLInk`,  you could use that to trigger another poll

Reply:
Gotcha. Would using [futures_timer::Delay](https://docs.rs/futures-timer/3.0.2/futures_timer/struct.Delay.html) a good approach too?

Reply:
Gotcha! Thanks! :)